### PR TITLE
PUBDEV-6938 TF-IDF

### PR DESCRIPTION
**Pull request by jancijen**
Implementation of measure TF-IDF used in information retrieval/text mining.

JIRA: https://0xdata.atlassian.net/browse/PUBDEV-6938

### CODE REVIEW

File: h2o-algos/src/test/java/hex/tfidf/TfIdfPreprocessorTest.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+public class TfIdfPreprocessorTest extends TestUtil {
+
+    @BeforeClass()
+    public static void setup() { stall_till_cloudsize(1); }
+
+    @Test
+    public void testTfIdfPreprocessing() {
+        Frame fr = getSimpleTestFrame();
+        byte[] outputTypes = new byte[]{Vec.T_STR, Vec.T_NUM, Vec.T_NUM};
```
Comment (maurever) [inline h2o-algos/src/test/java/hex/tfidf/TfIdfPreprocessorTest.java]: This line causes the error (`DistributedException from /192.168.0.164:54321: 'NewChunk has type String, but the Vec is of type Numeric'`). The third type (word tokens) should be `Vec.T_STR`.

Comment (jancijen) [reply h2o-algos/src/test/java/hex/tfidf/TfIdfPreprocessorTest.java]: Thank you. Careless mistake.

File: h2o-r/h2o-package/R/tf-idf.R [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+#' @param frame documents frame for which TF-IDF values should be computed.
+#'              Row format: documentID, documentString
+#'
+#' @return  resulting frame with TF-IDF values.
+#'          Row format: documentID, word, TF, IDF, TF-IDF
+h2o.tf_idf <- function(frame) {
+    if( is(frame, 'H2OFrame') ) {
+        .newExpr('tf-idf', frame)
+    } else {
+        warning(paste0("TF-IDF cannot be computed for class ", class(frame), ". Frame input is required."))
```
Comment (maurever) [inline h2o-r/h2o-package/R/tf-idf.R:12]:
```suggestion
        warning(paste0("TF-IDF cannot be computed for class ", class(frame), ". H2OFrame input is required."))
```

File: h2o-core/src/test/java/hex/tfidf/DocumentFrequencyTest.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            put("Z", 1L);
+        }};
+        int expectedDfValuesCnt = expectedDocumentFrequencies.size();
+
+        try {
+            DocumentFrequency df = new DocumentFrequency();
+            Frame outputFrame = df.compute(fr);
+            long outputRowsCnt = outputFrame.numRows();
+
+            Assert.assertEquals(expectedDfValuesCnt, outputRowsCnt);
```
Comment (maurever) [inline h2o-core/src/test/java/hex/tfidf/DocumentFrequencyTest.java:45]: Try to add messages to all asserts, please. :)

File: h2o-r/h2o-package/R/tf-idf.R (hunk 2 above)
Comment (jancijen) [reply h2o-r/h2o-package/R/tf-idf.R]: Good point. I changed it.

File: h2o-core/src/test/java/hex/tfidf/DocumentFrequencyTest.java (hunk 3 above)
Comment (jancijen) [reply h2o-core/src/test/java/hex/tfidf/DocumentFrequencyTest.java]: 👍 I added messages to all asserts.

File: h2o-algos/src/test/java/hex/tfidf/TfIdfTrainerTest.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                put("A", new int[]{1,3,0});
+                put("B", new int[]{1,0,1});
+                put("C", new int[]{1,0,3});
+                put("Z", new int[]{0,1,0});
+            }};
+
+            expWordsCounts.forEach((word, expCounts) -> {
+                Vec actualCounts = outputFrame.vec(wordsIndices.get(new BufferedString(word))._val);
+                for (int i = 0; i < expCounts.length; i++)
+                    assertEquals(expCounts[i], actualCounts.at8(i));
```
Comment (maurever) [inline h2o-algos/src/test/java/hex/tfidf/TfIdfTrainerTest.java:47]: Please, add a message to all asserts. :)

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+    @Override
+    public String str() {
+        return "tf-idf";
+    }
+
+    /**
+     * Final TF-IDF Map-Reduce task used to combine TF and IDF values together.
+     */
+    private static class TfIdf extends MRTask<TfIdf> {
```
Comment (maurever) [inline h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java:114]: I would recommend to name it as:
```suggestion
    private static class TfIdfTask extends MRTask<TfIdf> {
```

The same for your other MRTasks.

File: h2o-py/h2o/information_retrieval/tf_idf.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+                        </p>
+                        <p>
+                        Row format when <code>preprocess = False</code>: documentID, word
+                        </p>
+    :param preprocess:  whether input frame should be pre-processed. Defaults to <code>True</code>.
+
+    :return:    resulting frame with TF-IDF values. 
+                Row format: documentID, word, TF, IDF, TF-IDF
+    """
+
```
Comment (maurever) [inline h2o-py/h2o/information_retrieval/tf_idf.py:28]: Add check it is H2OFrame, please.

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    @Override
+    public String[] args() {
+        return new String[]{ "frame", "preprocess" };
+    }
+
+    @Override
+    public Val apply(Env env, Env.StackHelp stk, AstRoot[] asts) {
+        Frame inputFrame = stk.track(asts[1].exec(env).getFrame());
```
Comment (maurever) [inline h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java:47]: What if a user will put a frame with more than two columns?

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+            byte[] outputTypes = new byte[]{ Vec.T_NUM, Vec.T_STR };
+
+            TfIdfPreprocessor preprocessor = new TfIdfPreprocessor();
+            wordFrame = preprocessor.doAll(outputTypes, inputFrame).outputFrame(PREPROCESSED_FRAME_COL_NAMES, null);
+
+            documentsCnt = inputFrame.numRows();
+        } else {
+            wordFrame = inputFrame;
+            // TODO: Avoid converting to categorical Vec
+            documentsCnt = inputFrame.vec(0).toCategoricalVec().cardinality();
```
Comment (honzasterba) [inline h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java:63]: 1) AstUnique might by an alternative to converting to categorical
2) either or, you might want to track the newly created categorical vec so that it does not leak (I am not sure about this)

File: h2o-core/src/main/java/hex/tfidf/DocumentFrequency.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    /**
+     * Computes document frequency values for given words in documents.
+     *
+     * @param wordFrame input frame of words for which document frequency
+     *                  values should be computed. For exact format
+     *                  see {@link TermFrequency}.
+     *
+     * @return frame containing document frequency values for given words.
+     */
+    public Frame compute(Frame wordFrame) {
```
Comment (honzasterba) [inline h2o-core/src/main/java/hex/tfidf/DocumentFrequency.java:25]: can be static

File: h2o-core/src/test/java/hex/tfidf/DocumentFrequencyTest.java [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+        Set<String> doc1Tokens = new HashSet<>(tokens.subList(0, docSplitIndex));
+        Set<String> doc2Tokens = new HashSet<>(tokens.subList(docSplitIndex, tokensCnt));
+        
+        Map<String, Long> expectedDFs = doc1Tokens.stream().collect(Collectors.toMap(token -> token, token -> 1L));
+        doc2Tokens.forEach(token -> expectedDFs.merge(token, 1L, Long::sum));
+
+        System.out.println("Computing DF...");
+        try {
+            DocumentFrequency df = new DocumentFrequency();
```
Comment (honzasterba) [inline h2o-core/src/test/java/hex/tfidf/DocumentFrequencyTest.java:91]: unnecessary variable

File: h2o-core/src/test/java/hex/tfidf/DocumentFrequencyTest.java [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+                .withChunkLayout(chunkLayout.stream().mapToLong(Integer::intValue).toArray())
+                .build();
+
+        Set<String> doc1Tokens = new HashSet<>(tokens.subList(0, docSplitIndex));
+        Set<String> doc2Tokens = new HashSet<>(tokens.subList(docSplitIndex, tokensCnt));
+        
+        Map<String, Long> expectedDFs = doc1Tokens.stream().collect(Collectors.toMap(token -> token, token -> 1L));
+        doc2Tokens.forEach(token -> expectedDFs.merge(token, 1L, Long::sum));
+
+        System.out.println("Computing DF...");
```
Comment (honzasterba) [inline h2o-core/src/test/java/hex/tfidf/DocumentFrequencyTest.java:89]: use Logging and debug (Log4j)

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+        }
+
+        // DF
+        DocumentFrequency dfTaks = new DocumentFrequency();
```
Comment (honzasterba) [inline h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java:67]: unnecessary variable

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+        dfOutFrame.remove(dfOutFrame.numCols() - 1);
+        dfOutFrame.add(IDF_COL_NAME, idfValues);
+
+        // TF
+        TermFrequency tfTask = new TermFrequency();
+        Frame tfOutFrame = tfTask.compute(wordFrame);
+        
+        // Intermediate frame containing both TF and IDF values
+        // TODO: Avoid converting a string column to categorical
+        tfOutFrame.replace(1, tfOutFrame.vecs()[1].toCategoricalVec());
```
Comment (honzasterba) [inline h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java:83]: could use AstUnique here as well, also I see potential leaks all over this method

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+        // TF-IDF
+        int tfOutFrameColCnt = tfIdfIntermediate.numCols();
+        TfIdf tfIdfTask = new TfIdf(tfOutFrameColCnt - 2, tfOutFrameColCnt - 1);
+        Vec tfIdfValues = tfIdfTask.doAll(new byte[]{ Vec.T_NUM }, tfIdfIntermediate).outputFrame().anyVec();
+
+        // Construct final frame containing TF, IDF and TF-IDF values
+        tfIdfIntermediate.add(TF_IDF_COL_NAME, tfIdfValues);
+        tfIdfIntermediate._key = Key.make();
+
+        System.out.println(tfIdfIntermediate.toTwoDimTable().toString());
```
Comment (honzasterba) [inline h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java:101]: use logging ideally wrapped in isDebugEnabled

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+
+        /**
+         * Computes TF-IDF value from given TF and IDF values.
+         * 
+         * @param tf    TF value to be used to compute TF-IDF.
+         * @param idf   IDF value to be used to compute TF-IDF.
+         *
+         * @return TF-IDF value from given TF and IDF values.
+         */
+        private static double tfIdf(final long tf, final double idf) {
```
Comment (honzasterba) [inline h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java:149]: unnecessary function

File: h2o-algos/src/test/java/hex/tfidf/TfIdfTrainerTest.java (hunk 4 above)
Comment (jancijen) [reply h2o-algos/src/test/java/hex/tfidf/TfIdfTrainerTest.java]: Sorry, this test is not used anymore. I removed it.

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java (hunk 5 above)
Comment (jancijen) [reply h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java]: Sounds good. I renamed all of the `MRTask` in such way. Also, I renamed "tasks" which are not actual `MRTaks` - `DocumentFrequency`, `TermFrequency`... to be consistent.

File: h2o-py/h2o/information_retrieval/tf_idf.py (hunk 6 above)
Comment (jancijen) [reply h2o-py/h2o/information_retrieval/tf_idf.py:31]: Good point. Added.

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java (hunk 7 above)
Comment (jancijen) [reply h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java:62]: I added check for the input types - `Num`, `Str` and number of columns (`>= 2`). Or would you prefer to have parameters to indicate which columns should be used?

I also added corresponding unit tests for this, and added check for empty input.

File: h2o-algos/src/test/java/hex/tfidf/Word2IndexTest.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    @Test
+    public void testSmallDataWordsIndexing() {
+        Frame fr = getSimpleTestFrame();
+
+        try {
+            Word2Index word2index = new Word2Index(" ").doAll(fr.vec(0));
+
+            String[] expUniqueWords = new String[] { "A", "B", "C", "Z" };
+            for (String expWord : expUniqueWords)
+                Assert.assertTrue(word2index._wordsIndices.containsKey(new BufferedString(expWord)));
```
[the remark this answers is not in the document]
Comment (jancijen) [reply h2o-algos/src/test/java/hex/tfidf/Word2IndexTest.java]: +1: I changed it in all test classes.

File: h2o-core/src/main/java/hex/tfidf/DocumentFrequency.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+        
+        int[] groupByColumns = new int[]{ 1 };
+        
+        // Construct final frame with DF values
+        Frame dfFrame = new AstGroup().performGroupingWithAggregations(uniqueWordsPerDocFrame,
+                                                                       groupByColumns,
+                                                                       aggs).getFrame();
+        dfFrame.setNames(OUTPUT_FRAME_COL_NAMES);
+        
+        uniqueWordsPerDocFrame.remove();
```
[the remark this answers is not in the document]
Comment (jancijen) [reply h2o-core/src/main/java/hex/tfidf/DocumentFrequency.java]: Good point. I added `Scope` to handle it.

File: h2o-core/src/main/java/hex/tfidf/DocumentFrequency.java (hunk 9 above)
Comment (jancijen) [reply h2o-core/src/main/java/hex/tfidf/DocumentFrequency.java]: Good point. I changed it for both `TermFrequency` and `DocumentFrequency` and in places where `compute` method is being used.

File: h2o-core/src/test/java/hex/tfidf/DocumentFrequencyTest.java (hunk 10 above)
Comment (jancijen) [reply h2o-core/src/test/java/hex/tfidf/DocumentFrequencyTest.java]: Now solved by your previous suggestion (to make it static).

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java (hunk 12 above)
Comment (jancijen) [reply h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java]: Now solved by your previous suggestion (to make it static).

File: h2o-core/src/test/java/hex/tfidf/DocumentFrequencyTest.java (hunk 11 above)
Comment (jancijen) [reply h2o-core/src/test/java/hex/tfidf/DocumentFrequencyTest.java]: +1: Good point. I replaced standard output prints with debug logs throughout the code.

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java (hunk 14 above)
Comment (jancijen) [reply h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java]: 👍 Changed.

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java (hunk 15 above)
Comment (jancijen) [reply h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java]: This was there just for the code readability, but I guess in this case it is quite obvious and strictly defined to be a product. I removed it.

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java (hunk 8 above)
Comment (jancijen) [reply h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java]: Good idea. I replaced it with `AstUnique` and `Vec#length()`.

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java (hunk 13 above)
Comment (jancijen) [reply h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java]: In this case I need to merge `tfOutFrame` and `dfOutFrame`, where term frequency values are "per word and document" and document frequency values which are "per word". Therefore, I need to add one value from `dfOutFrame` to multiple rows in `tfOutFrame`. How would you apply `AstUnique` to solve this?

I checked and it seems that long time ago there has been an implementation of `AstMerge` that worked with string values - see `mergetest.py` in https://github.com/h2oai/h2o-3/commit/f645ee8c6d86749845d033c4d1e4851bb9c96d52, but this no longer works.

But for now, it seems that the main reason why I was worried about conversion to `Categorical` vector - the `Categorical.MAX_CATEGORICAL_COUNT` does not apply for `String` vectors that are being transformed to `Categorical` vector. I tried to convert `String` vector containing `Categorical.MAX_CATEGORICAL_COUNT + 10` unique values to `Categorical` vector and it worked just fine.

Is there a better solution for this?

Also, I fixed the leaks by using `Scope`.

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java (hunk 7 above)
Comment (maurever) [reply h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java:64]: I think, the parameter for indicating the columns should be the best way. WDYT @honzasterba  and @michalkurka ?

Comment (honzasterba) [reply h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java:64]: sounds good to me, something like this:

`tf_idf(frame, document_id=(str, int), document_word=(str,int), preprocess, case_sens)`

I think expecting that the frame has some pre-defined structure or column names is not something we do very often in H2O

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java (hunk 13 above)
Comment (honzasterba) [reply h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java]: Converting to categorical and doing merge is just hiding the problem. The problem is that you need to do a merge of two frames over a very large string-based key.
From my quick read of the code both tfOutFrame and dfOutFrame are computed from the same source wordsFrame, it might be possible to make this more efficient by combining both of the computations of tfOut and dfOut into a single series of MRTasks.
On the other hand how big can the domain really be? We are taking about text here, so the domains should not be bigger than couple 100k. So maybe the categorical+merge is not such a big deal.
@michalkurka any thoughts?

**Comment by maurever** [comment]
Mr. Jenkins, please test!

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java (hunk 13 above)
Comment (jancijen) [reply h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java]: I know. Well, yes that is true, both `tfOutFrame` and `dfOutFrame` are built by using same input, but it is done via `GroupBy`, where `TermFrequency` is done on grouping by `docID` and `word`, and `DocumentFrequency` first groups by the same columns (I changed it in the last commit to expecte unique words per document, therefore reuse the output of `TermFrequencyTask` since it is not a publicly available API and it saves some computation.) and then do another groupby operation in order to count number of documents containing given word.

Well, to be honest it looks like that even `Categorical.MAX_CATEGORICAL_COUNT` shouldn't be a problem even for larger document corpora. But if there is a better way to solve this I would prefer that over converting it to a categorical vector.

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java (hunk 7 above)
Comment (jancijen) [reply h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java:73]: I changed it to expect column indices for `AstTfIdf` and for `R` and `Python` methods both column names and indices are allowed.

**Comment by maurever** [comment]
So, we finished with @jancijen last updates of code to pass all tests. The last thing is failing and also fixed - tests are running right now (#4668 ). It is just a small issue in the Python 2.7 demo. So please, @honzasterba and @michalkurka, could you do a quick review again, that everything was implemented correctly, please? Thank you!

**Comment by maurever** [comment]
The tests in the #4668 PR are green. Thank you, @jancijen, for your contribution to the H2O-3 platform! Thanks, @honzasterba, for your review!
